packages/devtools_app/lib/src/inspector/inspector_service.dart

```diff
@@ -179,11 +201,115 @@ class InspectorService extends DisposableController
+  Set<String> _rootPackages;
+
+  @visibleForTesting
+  List<String> get rootPackagePrefixes => _rootPackagePrefixes;
+  List<String> _rootPackagePrefixes;
+
+  Future<void> _onRootDirectoriesChanged(List<String> directories) async {
+    _rootDirectories.value = directories;
+    _rootPackages = {};
+    _rootPackagePrefixes = [];
+    for (var directory in directories) {
+      // TODO(jacobr): add an API to DDS to provide the actual mapping to and
+      // from absolute file paths to packages instead of having to guess it
+      // here.
+      assert(!directory.startsWith('package:'));
+
+      final parts =
+          directory.split('/').where((element) => element.isNotEmpty).toList();
+      final libIndex = parts.lastIndexOf('lib');
+      final path = libIndex > 0 ? parts.sublist(0, libIndex) : parts;
+      // Special case handling of bazel packages.
+      final google3Index = path.lastIndexOf('google3');
+      if (google3Index != -1 && google3Index + 1 < path.length) {
+        var packageParts = path.sublist(google3Index + 1);
+        //
+        if (packageParts[0] == 'third_party' && packageParts.length >= 3) {
+          assert(packageParts[1] == 'dart' || packageParts[1] == 'dart_src');
+          packageParts = packageParts.sublist(2);
+        }
+        final google3PackageName = packageParts.join('.');
+        _rootPackages.add(google3PackageName);
+        _rootPackagePrefixes.add(google3PackageName + '.');
+      } else {
+        _rootPackages.add(path.last);
+      }
+    }
+
+    await _updateLocalClasses();
+  }
+
+  Future<void> _updateLocalClasses() async {
+    localClasses.clear();
+    if (_rootDirectories.value.isNotEmpty) {
+      final isolate = inspectorLibrary.isolate;
+      for (var libraryRef in isolate.libraries) {
+        if (isLocalUri(libraryRef.uri)) {
+          final Library library = await inspectorLibrary.service
+              .getObject(isolate.id, libraryRef.id);
+          for (var classRef in library.classes) {
+            localClasses[classRef.name] = classRef;
+          }
+        }
+      }
+    }
+  }
+
+  @visibleForTesting
+  bool isLocalUri(String rawUri) {
+    final uri = Uri.parse(rawUri);
+    if (uri.scheme != 'file' && uri.scheme != 'dart') {
+      // package scheme or some other dart specific scheme.
+      final packageName = uri.pathSegments.first;
+      if (_rootPackages.contains(packageName)) return true;
+
+      // This attempts to gracefully handle the bazel package case.
+      return _rootPackagePrefixes
+          .any((prefix) => packageName.startsWith(prefix));
+    }
+    for (var root in _rootDirectories.value) {
+      if (root.endsWith(rawUri)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @visibleForTesting
+  final Map<String, ClassRef> localClasses = {};
+
+  bool isLocalClass(RemoteDiagnosticsNode node) {
+    if (node.widgetRuntimeType == null) return false;
+    final rawType = node.widgetRuntimeType.split('<').first;
```

Comment:
add an example of what `node.widgetRuntimeType` should look like in a comment

Reply:
done.